src/linker/Linker/SerializationMarker.cs

```diff
@@ -0,0 +1,202 @@
+	// For each recursive type:
+	//   mark the type and its public instance {fields, properties, and parameterless constructors}
+
+	[Flags]
+	public enum SerializerKind
+	{
+		None = 0,
+		XmlSerializer = 1,
+		DataContractSerializer = 2,
+	}
+
+	public class SerializationMarker
+	{
+		readonly LinkContext _context;
+
+		SerializerKind ActiveSerializers { get; set; }
+
+		Dictionary<SerializerKind, HashSet<(TypeDefinition, ICustomAttributeProvider)>> _trackedTypes;
+		Dictionary<SerializerKind, HashSet<(TypeDefinition, ICustomAttributeProvider)>> TrackedTypes {
+			get {
+				if (_trackedTypes == null)
+					_trackedTypes = new Dictionary<SerializerKind, HashSet<(TypeDefinition, ICustomAttributeProvider)>> ();
+
+				return _trackedTypes;
+			}
+		}
+
+		HashSet<TypeDefinition> _recursiveTypes;
+		HashSet<TypeDefinition> RecursiveTypes {
+			get {
+				if (_recursiveTypes == null)
+					_recursiveTypes = new HashSet<TypeDefinition> ();
+
+				return _recursiveTypes;
+			}
+		}
+
+		public SerializationMarker (LinkContext context)
+		{
+			_context = context;
+		}
+
+		public bool IsActive (SerializerKind serializerKind) => ActiveSerializers.HasFlag (serializerKind);
+
+		static DependencyKind ToDependencyKind (SerializerKind serializerKind) => serializerKind switch {
+			SerializerKind.DataContractSerializer => DependencyKind.DataContractSerialized,
+			SerializerKind.XmlSerializer => DependencyKind.XmlSerialized,
+			_ => throw new ArgumentException (nameof (SerializerKind))
+		};
+
+		public void TrackForSerialization (TypeDefinition type, ICustomAttributeProvider provider, SerializerKind serializerKind)
+		{
+			if (ActiveSerializers.HasFlag (serializerKind)) {
+				MarkRecursiveMembers (type, new DependencyInfo (ToDependencyKind (serializerKind), provider));
+				return;
+			}
+
+			if (!TrackedTypes.TryGetValue (serializerKind, out var types)) {
+				types = new HashSet<(TypeDefinition, ICustomAttributeProvider)> ();
+				TrackedTypes.Add (serializerKind, types);
+			}
+
+			types.Add ((type, provider));
+		}
+
+		public void Activate (SerializerKind serializerKind)
+		{
+			if (!Enum.IsDefined<SerializerKind> (serializerKind) || serializerKind == SerializerKind.None)
+				throw new ArgumentException ($"Unexpected serializer kind {nameof (serializerKind)}");
+
+			if (ActiveSerializers.HasFlag (serializerKind))
+				return;
+
+			ActiveSerializers |= serializerKind;
+
+			if (!TrackedTypes.TryGetValue (serializerKind, out var types))
+				return;
+
+			foreach (var (type, provider) in types)
+				MarkRecursiveMembers (type, new DependencyInfo (ToDependencyKind (serializerKind), provider));
+
+			TrackedTypes.Remove (serializerKind);
+		}
+
+		public void MarkRecursiveMembers (TypeReference typeRef, in DependencyInfo reason)
+		{
+			MarkRecursiveMembersInternal (typeRef, reason);
+		}
+
+		void MarkRecursiveMembersInternal (TypeReference typeRef, in DependencyInfo reason)
```

Comment:
Why do we need the internal version if the public one is exactly the same thing?

Reply:
Removing (I think it was left over from a previous iteration that treated root types differently)

Reply:
Actually with the fix for https://github.com/mono/linker/pull/1932#discussion_r614053345 this will remain.